### ISSUE TITLE
Kubernetes API Authentication Proxy Request Headers

### ISSUE THREAD

**Issue by jontypreston** [kubernetes/kubernetes#82468]
**What would you like to be added**:
See #66754.

Assuming the `--requestheader-group-headers` flag is set to `X-Remote-Group` the API server should accept both the following header(s) to apply the groups A,B,C to the user:

```http
X-Remote-Group: A
X-Remote-Group: B
X-Remote-Group: C
```

```http
X-Remote-Group: A,B,C
```

Currently only the former is supported.

**Why is this needed**:
The HTTP spec ([RFC 2616](https://tools.ietf.org/html/rfc2616#section-4.2)) specifies how multiple headers of the same name can be combined into one.
> Multiple message-header fields with the same field-name MAY be
   present in a message if and only if the entire field-value for that
   header field is defined as a comma-separated list [i.e., #(values)].
   It MUST be possible to combine the multiple header fields into one
   "field-name: field-value" pair, without changing the semantics of the
   message, by appending each subsequent field-value to the first, each
   separated by a comma.

Popular tools, such as Python's WSGI, do not support sending multiple headers of the same name. This makes using a Python web server as an authenticating proxy troublesome. 

The previous issue (#66754) was closed as there's a chance some groups (eg from LDAP) could contain commas.

Kubernetes could explicitly forbid group names from containing commas.
Alternatively groups containing commas could be allowed if the commas are properly escaped in a similar way to how [LDAP supports commas in group names](https://www.ibm.com/support/knowledgecenter/SSPREK_9.0.0/com.ibm.isam.doc/base_admin/concept/con_validcharldapusrandgrpname.html).

**Comment by liggitt** [comment]
we may need a second named header to do this compatibly

**Comment by liggitt** [comment]
cc @mikedanese

**Comment by liggitt** [comment]
cc @deads2k

**Comment by deads2k** [comment]
I think we need to preserve the behavior that group names can contain commas.  I'm not sure this can be made backward compatible using a single field.

**Comment by mikedanese** [comment]
What does HTTP/2 say? This seems like an FR for go.

**Comment by liggitt** [comment]
> This seems like an FR for go.

not really... this is about how the `k8s.io/apiserver` requestheader proxy auth module treats headers on the incoming request. It does not treat commas as special characters for the multi-header attributes (groups and extra values), which the RFC apparently requires.

**Comment by sctlee** [comment]
Hi there, I try to support this in #83149. As mentioned above, this may have compatibility issues. 

I saw that someone provided an extra parameter `auth-header-groups-field-separator` to implemented this in [kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy). But I think that change is a bit too complicated.

**Comment by liggitt** [comment]
Also applies to impersonation headers

**Comment by travisgroth** [comment]
Thanks @liggitt 

Is there something we can do to move this forward?  I expected the comma support in groups might be a concern but that seems like quite the corner case.  Requiring `\` escaping seems very reasonable for that exceptional situation.  I don't think I've ever seen a group name with a comma in practice.

**Comment by liggitt** [comment]
> Is there something we can do to move this forward?

Starting a [proposal](https://github.com/kubernetes/enhancements/tree/master/keps) to describe the problem and propose 
 a solution that maintains backwards and forwards compatibility would be the next step.

> I expected the comma support in groups might be a concern but that seems like quite the corner case. I don't think I've ever seen a group name with a comma in practice.

LDAP groups are commonly mapped to group names, and if the fully-qualified name is used, almost always contain commas.

> Requiring `\` escaping seems very reasonable for that exceptional situation. 

Changing encoding of values in an existing API is a breaking change. Doing this compatibly would likely require a second set of headers (which adds its own compatibility issues around existing servers not expecting / honoring / filtering / appropriately protecting them)

**Comment by travisgroth** [comment]
> Changing encoding of values in an existing API is a breaking change.

In theory this could be solved using a non-standard separator and forcing the client to compile the header themselves.  `;` is a reasonable possibility.  But if _any_ change to the field semantics is considered breaking, that sounds like it isn't an option.

Is there a hard requirement for no breaking changes in the header spec?  The current behavior is in contradiction to [rfc2616](https://tools.ietf.org/html/rfc2616#section-4.2), as cited in the original report.  This should make it the current behavior a bug rather than an enhancement. 

Anyway, if breaking changes for groups with any possible separator in the name is not acceptable, we're stuck with new headers.  These would supersede the old ones and would follow the RFC for commas.  

Lacking any imagination, putting a version on the headers is probably the best thing to do for future changes:
`impersonate-group-v2`
`impersonate-user-v2`
 ...etc, would allow any special handling to fall under a wildcard (`impersonate-group*`), and provide a path for future changes.

Another potential format would be: `x-kube-auth-vYY-XXXX` with YY being version and XXXX being the header name such as `impersonate-group`.  You can still do wildcard matching and get to version the entire set of headers with a prefix match.  This also makes it a bit more clear what the intended consumer of the headers is.  Any special handling could probably be done on `x-kube-auth-*`.  

In either format:
 - Servers could continue honoring the old headers unless they receive the new equivalents
 - Clients which are not updated would continue to work with newer servers until the header version is deprecated
 - Clients which exceed the server API version **and** use the new headers would still break, which I think is inevitable

Would that be a reasonable proposal to submit?

**Comment by liggitt** [comment]
> In theory this could be solved using a non-standard separator and forcing the client to compile the header themselves. `;` is a reasonable possibility. But if _any_ change to the field semantics is considered breaking, that sounds like it isn't an option.
> 
> Is there a hard requirement for no breaking changes in the header spec? The current behavior is in contradiction to [rfc2616](https://tools.ietf.org/html/rfc2616#section-4.2), as cited in the original report. This should make it the current behavior a bug.

It is a bug, but since clients which can construct multiple headers can communicate values containing commas to the server today, and we don't reserve or escape characters in the group names, and the consequences of changing how authentication headers are interpreted are so severe, we would not change our interpretation of the headers sent by the client without an opt-in of some sort.

Rather than add new headers containing group names that need to be protected by auth proxies, my first thought is to have an opt-in way for a client to indicate it is sending comma-separated group names or extra attributes, and then require literal commas in group names be `\-escaped` or something. As a strawman:

`X-Remote-Group-Comma-Separated: true`

Obviously, that would only be honored by a new kubernetes API server, but if such a client cannot send multiple identical headers, I don't see a way for it to work against an older kubernetes API server anyway.

We would have to consider the implications of skew in both directions:

1. new kubernetes API server + old proxy (unaware of the opt-in header, blindly forwarding a user-specified `X-Remote-Group-Comma-Separated` header), where the old proxy:
    A. sets one or more group headers not containing commas - safe
    B. sets one or more group headers containing commas - a user could specify `X-Remote-Group-Comma-Separated: true` and the proxy could forward it, resulting in different and potentially dangerous group membership for the user; if the user was in `department=accounting,system:masters`, the user could make Kubernetes interpret their group membership as `department=accounting`,`system:masters`, which would make them a superuser.
2. old kubernetes API server + new proxy (unable to send multiple headers, attempting to set `X-Remote-Group-Comma-Separated: true`), where the new proxy:
    A. sets a group header not containing commas - safe
    B. sets a group header of `group1,group2` intending to be interpreted as `group1`,`group2` - the kubernetes API server will treat the user as having a single group of `group1,group2`

1B seems like it has the biggest security risk, since components of an existing group name could plausibly have higher permissions attached (imagine `O=Acme,OU=Accounting,Administrators` getting interpreted as [`O=Acme`,`OU=Accounting`,`Administrators`] instead)

2B is really only a risk if the proxy isn't sure whether the backing Kubernetes API server supports the comma-separated header.

Guidance for new proxies would be to disallow forwarding of a `X-Remote-Group-Comma-Separated` header from a user to be sure the groups would be interpreted as intended.

**Comment by enj** [comment]
If we would need to define a new header anyway, why not create a header like `X-Remote-Kubernetes-User-Info` and have the value be a JSON marshaled (and possibly base64 encoded) struct:

https://github.com/kubernetes/kubernetes/blob/4425fd6bfeef7f433c79e3f50e292a11681436cc/staging/src/k8s.io/api/authentication/v1/types.go#L102-L119

This would eliminate the need for multiple headers all together and would allow us to manage this data as a proper API.

**Comment by DocX** [comment]
This is also issue with HTTP client libraries in `ruby` that does not support sending multiple request headers with same name. It is of course possible to send it on IO level but that is impractical.

Same goes for the `Impersonate-User` header

**Comment by lmktfy** [comment]
If base64 encoding a value, I might use CBOR. It's already not human readable as base64, and CBOR is more compact.
